### PR TITLE
hashing python

### PR DESCRIPTION
**Pull request by SansPapyrus683**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as
done._

- [x] I have tested my code
- [x] I have followed the code style guidelines mentioned
      [here](https://usaco.guide/general/adding-solution/#code-conventions)
- [x] I have properly formatted the files according to the steps mentioned
      [here](https://usaco.guide/general/adding-solution#steps)

thinking of rewriting the solutions but that would take way too long given my iq

### CODE REVIEW

**Review by ryanchou-dev — APPROVED** [review]
looks good

**Review by TruaShamu — APPROVED** [review]
looks good to me

File: content/4_Gold/String_Hashing.mdx [hunk 1]
```diff
@@ -42,51 +42,115 @@ inverses.
...
+<PySection>
+
+```py
+class HashedString:
+	# Change M and P if you want
+	M = int(1e9) + 9
+	P = 9973
+
+	# pow[i] contains P^i % M
+	pow = [1]
```
Comment (Aadit-Ambadkar) [inline content/4_Gold/String_Hashing.mdx:133]:
```suggestion
	_pow = [1]
```
According to PEP 8 (https://pep8.org/#descriptive-naming-styles) use an underscore for weak private.

File: content/4_Gold/String_Hashing.mdx [hunk 2]
```diff
@@ -42,51 +42,115 @@ inverses.
...
+
+	def __init__(self, s: str):
+		while len(self.pow) < len(s):
+			self.pow.append((self.pow[-1] * self.P) % self.M)
```
Comment (Aadit-Ambadkar) [inline content/4_Gold/String_Hashing.mdx:137]:
```suggestion
		while len(self._pow) < len(s):
			self._pow.append((self._pow[-1] * self.P) % self.M)
```

File: content/4_Gold/String_Hashing.mdx [hunk 3]
```diff
@@ -42,51 +42,115 @@ inverses.
...
+
+		# p_hash[i] is the hash of the first i characters of the given string
+		self.p_hash = [0 for _ in range(len(s) + 1)]
```
Comment (Aadit-Ambadkar) [inline content/4_Gold/String_Hashing.mdx:140]:
```suggestion
		self._p_hash = [0 for _ in range(len(s) + 1)]
```
Again pep8

File: content/4_Gold/String_Hashing.mdx [hunk 4]
```diff
@@ -42,51 +42,115 @@ inverses.
...
+		for i in range(len(s)):
+			self.p_hash[i + 1] = (
+				((self.p_hash[i] * self.P) % self.M + ord(s[i]))
+			) % self.M
```
Comment (Aadit-Ambadkar) [inline content/4_Gold/String_Hashing.mdx:144]:
```suggestion
		for i in range(len(s)):
			self._p_hash[i + 1] = (
				((self._p_hash[i] * self.P) % self.M + ord(s[i]))
			) % self.M
```

File: content/4_Gold/String_Hashing.mdx [hunk 5]
```diff
@@ -42,51 +42,115 @@ inverses.
...
+
+	def get_hash(self, start: int, end: int) -> int:
+		raw_val = (
+			self.p_hash[end + 1] - (self.p_hash[start] * self.pow[end - start + 1])
+		)
+		return raw_val % self.M
```
Comment (Aadit-Ambadkar) [inline content/4_Gold/String_Hashing.mdx:150]:
```suggestion
	def get_hash(self, start: int, end: int) -> int:
		raw_val = (
			self._p_hash[end + 1] - (self._p_hash[start] * self._pow[end - start + 1])
		)
		return raw_val % self.M
```

**Review by Aadit-Ambadkar — CHANGES_REQUESTED** [review]

File: content/4_Gold/String_Hashing.mdx (hunk 2 above)
Comment (SansPapyrus683) [reply content/4_Gold/String_Hashing.mdx:137]: no need, i can refactor on my own

**Review by Aadit-Ambadkar — APPROVED** [review]
Looks good.

**Comment by SansPapyrus683** [comment]
3 content authors alr

**Pull request merged**